src/events/base.py

```diff
@@ -0,0 +1,74 @@
+#!/usr/bin/env python3
+# Copyright 2024 Canonical Limited
+# See LICENSE file for licensing details.
+
+"""Base utilities exposing common functionalities for all Events classes."""
+
+from functools import wraps
+from typing import Callable
+
+from ops import CharmBase, EventBase, Object, StatusBase
+
+from core.context import AuthProxyConfig, Context, IngressUrl, S3ConnectionInfo, Status
```

Comment:
Uhm, that's a small (but maybe relevant) point: here we could centralize external dependencies (charmlibs) into one module (e.g. core) instead of having the external imports on all modules. This has some, but not critical, advantages, like:

1. having a single proxy module where external dependencies are centralized 
2. Possibility of enhance external classes with extra capabilities, or patch behaviour. 

It is a pattern I kinda like, but up to you. I don't have strong opinion